### ISSUE TITLE
Potential size savings with mips16e instructions

### ISSUE THREAD

**Issue by eliminmax** [eliminmax/tiny-clear-elf#17]
Much like 32-bit ARM's Thumb instructions and RISC V's C extension, MIPS has shortened alternate instruction encodings available. If it's possible to use them in Tiny Clear ELF, they could cut the number of bytes by up to 18 (if all instructions used are MicroMIPS compatible). If I also address #13 for the `mipsel` and `mips64el` implementations, it could save even more.

**Comment by eliminmax** [comment]
From [the product page](https://mips.com/products/architectures/micromips/):

> microMIPS is supported in releases r3, r5 and r6 of the MIPS architecture

Given that Debian's minimum support version is r2, this does not seem to be an option.

That said, while researching this, I found out that `MIPS16e`, an older alternate encoding which microMIPS was designed to replace, is even more similar to Thumb instructions, in that they can't be freely mixed with full-length instructions, and you access them with odd addresses.

Unfortunately, I can't find any documentation stating definitively whether or not it's allowed to be used within Debian's `mipsel` binaries, or whether it's supported by all hardware Debian supports as part of `mipsel`. ~After downloading the netinst ISO for the `mipsel` architecture and disassembling almost all\* of the ELF executables in the included deb files, I found no 2-byte instructions.~


\* If multiple ELF executables had the same name, because I moved them into a single directory without protecting against clobbering, I only looked at one of them, but I still believe it's a representative sample. Also, not all ELF files would have been marked as executable, but still, I think that's enough data.

Then again, the same is true in `armel` - no 2-byte instructions were included in its ISO's executable elf files - and this time, I didn't clobber any of them. It's not proof that mips16e instructions are forbidden.
